core/trino-main/src/main/java/io/trino/operator/output/BlockTypeDispatchingPositionsAppender.java

```diff
@@ -0,0 +1,119 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Dispatches the {@link #append} and {@link #appendRle} methods to the {@link #delegate} depending on the input {@link Block} class.
+ * The {@link Block} is flattened if necessary so that the {@link #delegate} {@link PositionsAppender#append(IntArrayList, Block)}
+ * always gets flat {@link Block} and {@link PositionsAppender#appendRle(RunLengthEncodedBlock)} always gets {@link RunLengthEncodedBlock}
+ * with {@link RunLengthEncodedBlock#getValue()} being flat {@link Block}.
+ */
+public class BlockTypeDispatchingPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockTypeDispatchingPositionsAppender.class).instanceSize();
+
+    private final PositionsAppender delegate;
+
+    public BlockTypeDispatchingPositionsAppender(PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        if (source instanceof RunLengthEncodedBlock) {
+            delegate.appendRle(flatten((RunLengthEncodedBlock) source, positions.size()));
+        }
+        else if (source instanceof DictionaryBlock) {
+            appendDictionary(positions, (DictionaryBlock) source);
+        }
+        else {
+            delegate.append(positions, source);
+        }
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock source)
+    {
+        if (source.getPositionCount() == 0) {
+            return;
+        }
+        delegate.appendRle(flatten(source, source.getPositionCount()));
+    }
+
+    @Override
+    public Block build()
+    {
+        return delegate.build();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + delegate.getRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return delegate.getSizeInBytes();
+    }
+
+    private void appendDictionary(IntArrayList positions, DictionaryBlock source)
+    {
+        Block dictionary = source.getDictionary();
+
+        while (dictionary instanceof RunLengthEncodedBlock || dictionary instanceof DictionaryBlock) {
+            if (dictionary instanceof RunLengthEncodedBlock) {
+                // since at some level, dictionary contains only a single value, it can be flattened to rle
```

Comment:
`since at some level, dictionary contains only a single value`
->
`if at some level dictionary contains only a single value then it can be flattened to rle`